core/src/main/java/org/infinispan/expiration/impl/ClusterExpirationManager.java

```diff
@@ -0,0 +1,114 @@
+import org.infinispan.context.InvocationContextFactory;
+import org.infinispan.factories.KnownComponentNames;
+import org.infinispan.factories.annotations.ComponentName;
+import org.infinispan.factories.annotations.Inject;
+import org.infinispan.interceptors.InterceptorChain;
+
+import java.util.Iterator;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+
+@ThreadSafe
+public class ClusterExpirationManager<K, V> extends ExpirationManagerImpl<K, V> {
+
+   private ExecutorService asyncExecutor;
+   private AdvancedCache<K, V> cache;
+
+   @Inject
+   public void inject(AdvancedCache<K, V> cache,
+           @ComponentName(KnownComponentNames.ASYNC_OPERATIONS_EXECUTOR) ExecutorService asyncExecutor) {
+      this.cache = cache;
+      this.asyncExecutor = asyncExecutor;
+   }
+
+   @Override
+   public void processExpiration() {
+      long start = 0;
+      if (!Thread.currentThread().isInterrupted()) {
+         try {
+            if (trace) {
+               log.trace("Purging data container of expired entries");
+               start = timeService.time();
+            }
+            long currentTimeMillis = timeService.wallClockTime();
+            for (Iterator<InternalCacheEntry<K, V>> purgeCandidates = dataContainer.iteratorIncludingExpired();
+                 purgeCandidates.hasNext();) {
+               InternalCacheEntry<K, V> e = purgeCandidates.next();
+               if (e.canExpire()) {
+                  if (ExpiryHelper.isExpiredMortal(e.getLifespan(), e.getCreated(), currentTimeMillis)) {
+                     handleLifespanExpireEntry(e);
+                  } else if (ExpiryHelper.isExpiredTransient(e.getMaxIdle(), e.getLastUsed(), currentTimeMillis)) {
+                     super.handleInMemoryExpiration(e);
+                  }
+               }
+            }
+            if (trace) {
+               log.tracef("Purging data container completed in %s",
+                       Util.prettyPrintTime(timeService.timeDuration(start, TimeUnit.MILLISECONDS)));
+            }
+         } catch (Exception e) {
+            log.exceptionPurgingDataContainer(e);
+         }
+      }
+
+      if (!Thread.currentThread().isInterrupted()) {
+         persistenceManager.purgeExpired();
+      }
+   }
+
+   void handleLifespanExpireEntry(InternalCacheEntry<K, V> entry) {
+      K key = entry.getKey();
+      // We remove the entry if it is the same from the datacontainer while also submitting an async remove expired
+      // to other members of the cluster
+      // The most used case will be a miss so no extra read before
+      if (expiring.putIfAbsent(key, key) == null) {
+         long lifespan = entry.getLifespan();
+         if (trace) {
+            log.tracef("Submitting expiration removal for key %s which had lifespan of %s", key, lifespan);
+         }
+         asyncExecutor.submit(() -> {
```

Comment:
Ok, here's how I think this could have been simplified:

In the inject method, construct a ReadWriteMap out of a AdvancedCache (via a FunctionalMapImpl), and here, call `eval(K, V, BiFunction<V, ReadWriteEntryView<K, V>, R> f)` passing in the key, value, and a BiFunction implementation that takes lifespan in constructor. In the BiFunction apply implementation, extract find lifespan metaparam (this part I'm not sure if it's in place, in terms of interoperability with data stored via normal cache impl, but should be easy to put in place), then compare both lifespan and the value, and remove if needed.

On top of that, since eval() is async already, no need to depend on the async executor.

I agree that all pieces might not be in place to achieve this (the only one I have doubt about is whether you can retrieve lifespan from metadata for data stored via Cache/AdvancedCache), but it's really were we need to move towards, because as I said in the blog, we have tons of redundancy internally.